lib/Conversion/StableHLOToTTIR/StableHLOToTTIRPatterns.cpp

```diff
@@ -2603,162 +2273,510 @@ class StableHLOToTTIRLogicalAndBitwiseOpConversionPattern
+    } else {
+      rewriter.replaceOpWithNewOp<BitwiseDestOp>(srcOp, outputType,
+                                                 adaptor.getOperands());
+    }
+
+    return success();
+  }
+
+private:
+  enum StableHLOOpType { kLogical = 0, kBitwise = 1 };
+
+  // Determines stablehlo op type based on its operand types (i.e. their
+  // bit width). This assumes boolean operands are modeled as 1bit wide ints.
+  static StableHLOOpType getStableHLOOpType(const SrcOp &srcOp) {
+    // Checks if all operands are boolean (have bit width equal to 1).
+    bool allOperandsAreBoolean =
+        llvm::all_of(srcOp->getOperands(), [](auto operand) {
+          return mlir::cast<RankedTensorType>(operand.getType())
+                     .getElementTypeBitWidth() == 1;
+        });
+
+    return allOperandsAreBoolean ? StableHLOOpType::kLogical
+                                 : StableHLOOpType::kBitwise;
+  }
+
+  // This function iterates over each operand of the provided source operation
+  // and determines if each operand represents a boolean value through a
+  // stablehlo::ConvertOp from a boolean tensor.
+  // return true if all operands of a given StableHLO operation are effectively
+  // boolean.
+  bool areOperandsBoolean(const SrcOp &srcOp) const {
+    bool allOperandsBoolean =
+        llvm::all_of(srcOp->getOperands(), [](auto operand) {
+          // Check if operand is a ranked tensor of bit width 1.
+          if (mlir::cast<RankedTensorType>(operand.getType())
+                  .getElementTypeBitWidth() == 1) {
+            return false; // Already boolean, so not counted as converted.
+          }
+
+          // Check if the defining op is a stablehlo::ConvertOp.
+          auto definingOp = operand.getDefiningOp();
+          if (!isa_and_nonnull<mlir::stablehlo::ConvertOp>(definingOp)) {
+            return false; // Not a conversion from boolean
+          }
+
+          // Check if the input to the conversion is boolean.
+          return mlir::cast<RankedTensorType>(
+                     definingOp->getOperand(0).getType())
+                     .getElementTypeBitWidth() == 1;
+        });
+
+    return allOperandsBoolean;
+  }
+};
+} // namespace
+
+static llvm::ErrorOr<ttcore::ReduceType>
+getReduceTypeFromRegion(Region &region) {
+  // TODO(wooseoklee): This pattern matching mechanism may need to be updated as
+  // we see complicated patterns of reduce block in the future.
+
+  assert(region.getBlocks().size() == 1 &&
+         "Region should have exactly one block");
+
+  // Add, Prod, Max, Min are the only supported reduce types for now.
+  // Invalid is default reduction type in TT-Metal, used when we do not do
+  // reduction, only copy source to output.
+  auto &block = region.front();
+  for (Operation &op : block) {
+    if (isa<mlir::stablehlo::AddOp>(op)) {
+      return ttcore::ReduceType::Sum;
+    } else if (isa<mlir::stablehlo::MulOp>(op)) {
```

Comment:
:warning: **llvm\-else\-after\-return** :warning:
do not use `` else `` after `` return ``
```suggestion
    } if (isa<mlir::stablehlo::MulOp>(op)) {
```